File: conversion/layers/resource/placement.py
# @author: Davy Bellens

"""Geometric Constraint and Geometry schemas: points, directions, placements."""

from conversion.layers.resource.math3d import (
    IDENTITY, cross, dot, mat_multiply, normalize, scale, subtract
)


def read_point(entity):
    coords = list(entity.Coordinates)
    while len(coords) < 3:
        coords.append(0.0)
    return tuple(float(c) for c in coords[:3])


def read_direction(entity, default=(0.0, 0.0, 1.0)):
    if entity is None:
        return default
    ratios = list(entity.DirectionRatios)
    while len(ratios) < 3:
        ratios.append(0.0)
    return tuple(float(r) for r in ratios[:3])


def axis_placement_matrix(placement):
    """IfcAxis2Placement2D or 3D -> 4x4 matrix."""
    if placement is None:
        return IDENTITY

    origin = read_point(placement.Location)

    if placement.is_a("IfcAxis2Placement2D"):
        ref = read_direction(getattr(placement, "RefDirection", None), (1.0, 0.0, 0.0))
        x_axis = normalize((ref[0], ref[1], 0.0))
        y_axis = (-x_axis[1], x_axis[0], 0.0)
        z_axis = (0.0, 0.0, 1.0)
    else:
        z_axis = normalize(read_direction(getattr(placement, "Axis", None), (0.0, 0.0, 1.0)))
        ref = read_direction(getattr(placement, "RefDirection", None), (1.0, 0.0, 0.0))
        projected = subtract(ref, scale(z_axis, dot(ref, z_axis)))
        if dot(projected, projected) < 1e-20:
            fallback = (1.0, 0.0, 0.0) if abs(z_axis[0]) < 0.9 else (0.0, 1.0, 0.0)
            projected = subtract(fallback, scale(z_axis, dot(fallback, z_axis)))
        x_axis = normalize(projected)
        y_axis = cross(z_axis, x_axis)

    return (
        (x_axis[0], y_axis[0], z_axis[0], origin[0]),
        (x_axis[1], y_axis[1], z_axis[1], origin[1]),
        (x_axis[2], y_axis[2], z_axis[2], origin[2]),
        (0.0, 0.0, 0.0, 1.0),
    )


def local_placement_matrix(placement):
    """Resolve an IfcLocalPlacement chain up to the world."""
    if placement is None:
        return IDENTITY
    if not placement.is_a("IfcLocalPlacement"):
        return IDENTITY

    local = axis_placement_matrix(placement.RelativePlacement)
    parent = local_placement_matrix(placement.PlacementRelTo)
    return mat_multiply(parent, local)

File: UnityPackage/com.byvad.ifc-import/Runtime/Layers/Resource/Placement.cs
// @author: Davy Bellens

using Conversion.Ifc;

namespace Conversion.Layers.Resource
{
    /// <summary>
    /// Geometric Constraint and Geometry schemas: points, directions, placements.
    /// </summary>
    public static class Placement
    {
        public static Vec3 ReadPoint(IfcEntity entity)
        {
            if (entity == null)
            {
                return Vec3.Zero;
            }
            double[] coordinates = entity.Doubles("Coordinates");
            return new Vec3(
                coordinates.Length > 0 ? coordinates[0] : 0.0,
                coordinates.Length > 1 ? coordinates[1] : 0.0,
                coordinates.Length > 2 ? coordinates[2] : 0.0);
        }

        public static Vec3 ReadDirection(IfcEntity entity, Vec3 fallback)
        {
            if (entity == null)
            {
                return fallback;
            }
            double[] ratios = entity.Doubles("DirectionRatios");
            if (ratios.Length == 0)
            {
                return fallback;
            }
            return new Vec3(
                ratios[0],
                ratios.Length > 1 ? ratios[1] : 0.0,
                ratios.Length > 2 ? ratios[2] : 0.0);
        }

        /// <summary>IfcAxis2Placement2D or 3D to a 4x4 transform.</summary>
        public static Matrix4 AxisPlacementMatrix(IfcEntity placement)
        {
            if (placement == null)
            {
                return Matrix4.Identity;
            }

            Vec3 origin = ReadPoint(placement.Entity("Location"));

            Vec3 xAxis, yAxis, zAxis;

            if (placement.IsA("IfcAxis2Placement2D"))
            {
                Vec3 reference = ReadDirection(placement.Entity("RefDirection"), Vec3.UnitX);
                if (!new Vec3(reference.X, reference.Y, 0.0).TryNormalize(out xAxis))
                {
                    xAxis = Vec3.UnitX;
                }
                yAxis = new Vec3(-xAxis.Y, xAxis.X, 0.0);
                zAxis = Vec3.UnitZ;
            }
            else
            {
                (xAxis, zAxis) = OrthonormalBasis(
                    ReadDirection(placement.Entity("Axis"), Vec3.UnitZ),
                    ReadDirection(placement.Entity("RefDirection"), Vec3.UnitX));
                yAxis = Vec3.Cross(zAxis, xAxis);
            }

            return Matrix4.FromBasis(xAxis, yAxis, zAxis, origin);
        }

        /// <summary>
        /// Build an orthonormal (xAxis, zAxis) pair from an approximate primary
        /// direction and a reference direction, Gram-Schmidt style. If the
        /// reference is parallel to the primary axis (or absent), falls back to
        /// any perpendicular rather than dividing by zero.
        /// </summary>
        private static (Vec3 XAxis, Vec3 ZAxis) OrthonormalBasis(Vec3 rawPrimary, Vec3 rawReference)
        {
            if (!rawPrimary.TryNormalize(out Vec3 zAxis))
            {
                zAxis = Vec3.UnitZ;
            }

            Vec3 projected = rawReference - zAxis * Vec3.Dot(rawReference, zAxis);
            if (projected.LengthSquared < 1e-20)
            {
                // The reference was parallel to the axis, or absent. Pick any
                // perpendicular rather than dividing by zero.
                Vec3 candidate = System.Math.Abs(zAxis.X) < 0.9 ? Vec3.UnitX : Vec3.UnitY;
                projected = candidate - zAxis * Vec3.Dot(candidate, zAxis);
            }

            if (!projected.TryNormalize(out Vec3 xAxis))
            {
                xAxis = Vec3.UnitX;
            }
            return (xAxis, zAxis);
        }

        /// <summary>Guards against a cyclic PlacementRelTo, not a limit any real chain should reach.</summary>
        private const int MaxPlacementChainLength = 256;

        /// <summary>
        /// Walk an IfcLocalPlacement chain up to the world.
        /// <para>
        /// Iterative rather than recursive: a deeply nested spatial hierarchy is a
        /// chain, not a tree, and it costs nothing to avoid the stack.
        /// </para>
        /// </summary>
        public static Matrix4 LocalPlacementMatrix(IfcEntity placement)
        {
            if (placement == null || !placement.IsA("IfcLocalPlacement"))
            {
                return Matrix4.Identity;
            }

            var chain = new System.Collections.Generic.List<IfcEntity>();
            for (IfcEntity step = placement;
                 step != null && step.IsA("IfcLocalPlacement");
                 step = step.Entity("PlacementRelTo"))
            {
                chain.Add(step);
                if (chain.Count > MaxPlacementChainLength)
                {
                    break;   // cyclic PlacementRelTo; malformed, but do not hang
                }
            }

            Matrix4 result = Matrix4.Identity;
            for (int i = chain.Count - 1; i >= 0; i--)
            {
                Matrix4 local = AxisPlacementMatrix(chain[i].Entity("RelativePlacement"));
                result = Matrix4.Multiply(result, local);
            }
            return result;
        }

        /// <summary>
        /// IfcCartesianTransformationOperator3D, as used by IfcMappedItem.MappingTarget.
        /// <para>
        /// Honours Scale2 and Scale3 when the operator is the nonUniform variant.
        /// Axis2 is re-derived rather than taken as authored: EXPRESS defines the
        /// operator's basis as an orthogonal right-handed set built from Axis1 and
        /// Axis3, so a supplied Axis2 is a hint to orthogonalise against, not a
        /// licence to shear.
        /// </para>
        /// </summary>
        public static Matrix4 TransformationOperatorMatrix(IfcEntity op)
        {
            if (op == null)
            {
                return Matrix4.Identity;
            }

            Vec3 origin = ReadPoint(op.Entity("LocalOrigin"));

            bool nonUniform = op.IsA("IfcCartesianTransformationOperator3DnonUniform");
            double scale1 = op["Scale"].IsNull ? 1.0 : op.Double("Scale", 1.0);
            double scale2 = nonUniform && !op["Scale2"].IsNull ? op.Double("Scale2", scale1) : scale1;
            double scale3 = nonUniform && !op["Scale3"].IsNull ? op.Double("Scale3", scale1) : scale1;

            (Vec3 xAxis, Vec3 zAxis) = OrthonormalBasis(
                ReadDirection(op.Entity("Axis3"), Vec3.UnitZ),
                ReadDirection(op.Entity("Axis1"), Vec3.UnitX));

            Vec3 yAxis = Vec3.Cross(zAxis, xAxis);

            // Axis2, when supplied, only decides which of the two perpendiculars
            // to prefer. Flip rather than replace, so the basis stays orthonormal.
            IfcEntity axis2 = op.Entity("Axis2");
            if (axis2 != null)
            {
                Vec3 hint = ReadDirection(axis2, yAxis);
                if (Vec3.Dot(hint, yAxis) < 0.0)
                {
                    yAxis = -yAxis;
                    zAxis = -zAxis;
                }
            }

            return Matrix4.FromBasis(xAxis * scale1, yAxis * scale2, zAxis * scale3, origin);
        }
    }
}